### PR TITLE
Add neighbor handling to interface handler

### PR DESCRIPTION
**Pull request by DanG100**

### CODE REVIEW

File: dataplane/forwarding/protocol/arp/arp.go [hunk 1]
```diff
@@ -115,8 +115,8 @@ func (ARP) Modify(fwdpb.PacketHeaderId) error {
 
 // parse parses an ARP header.
 func parse(frame *frame.Frame, desc *protocol.Desc) (protocol.Handler, fwdpb.PacketHeaderId, error) {
-	if frame.Len() != arpBytes {
-		return nil, fwdpb.PacketHeaderId_PACKET_HEADER_ID_NONE, fmt.Errorf("arp: parse failed, invalid frame length %v, expected %v bytes", frame.Len(), arpBytes)
+	if frame.Len() < arpBytes {
+		return nil, fwdpb.PacketHeaderId_PACKET_HEADER_ID_NONE, fmt.Errorf("arp: parse failed, invalid frame length %v, expected < %v bytes", frame.Len(), arpBytes)
```
[the remark this answers is not in the document]
Comment (DanG100) [reply dataplane/forwarding/protocol/arp/arp.go:119]: ethernet frames have a minimum size and may zero padded to fit that minimum size

File: dataplane/handlers/interface.go [hunk 2]
```diff
@@ -289,6 +299,66 @@ func (ni *Interface) handleAddrUpdate(ctx context.Context, au *netlink.AddrUpdat
...
 }
 
+// handleNeighborUpdate modifies the state based on changes to the neighbor.
+func (ni *Interface) handleNeighborUpdate(ctx context.Context, nu *netlink.NeighUpdate) {
+	ni.stateMu.Lock()
+	defer ni.stateMu.Unlock()
+	name := ni.idxToName[nu.LinkIndex]
+	if !engine.IsTap(name) {
+		return
+	}
```
[the remark this answers is not in the document]
Comment (DanG100) [reply dataplane/handlers/interface.go:309]: arp packets are forwarded to tap interface so they'll have neighbors

File: dataplane/handlers/interface.go [hunk 3]
```diff
@@ -289,6 +299,66 @@ func (ni *Interface) handleAddrUpdate(ctx context.Context, au *netlink.AddrUpdat
...
+			return
+		}
+		if nu.Family == unix.AF_INET6 {
+			neigh := sub.GetOrCreateIpv6().GetOrCreateNeighbor(nu.IP.String())
+			neigh.LinkLayerAddress = ygot.String(nu.HardwareAddr.String())
+			if nu.Flags&unix.NUD_PERMANENT != 0 {
+				neigh.Origin = oc.IfIp_NeighborOrigin_STATIC
+			} else {
+				neigh.Origin = oc.IfIp_NeighborOrigin_DYNAMIC
+			}
```
[the remark this answers is not in the document]
Comment (DanG100) [reply dataplane/handlers/interface.go:339]: always true

File: dataplane/internal/engine/table.go [hunk 4]
```diff
@@ -218,6 +220,83 @@ func createFIBSelector(ctx context.Context, c fwdpb.ServiceClient) error {
...
+// createLayer2PuntTable creates a table to packets to punt at layer 2 (input port and mac dst).
+func createLayer2PuntTable(ctx context.Context, c fwdpb.ServiceClient) error {
+	multicast := &fwdpb.TableCreateRequest{
+		ContextId: &fwdpb.ContextId{Id: contextID},
+		Desc: &fwdpb.TableDesc{
+			TableType: fwdpb.TableType_TABLE_TYPE_PREFIX,
+			TableId:   &fwdpb.TableId{ObjectId: &fwdpb.ObjectId{Id: layer2PuntTable}},
+			Actions:   []*fwdpb.ActionDesc{{ActionType: fwdpb.ActionType_ACTION_TYPE_CONTINUE}},
+			Table: &fwdpb.TableDesc_Prefix{
+				Prefix: &fwdpb.PrefixTableDesc{
```
[the remark this answers is not in the document]
Comment (DanG100) [reply dataplane/internal/engine/table.go:232]: multicast macs are ranges

File: dataplane/handlers/interface.go [hunk 5]
```diff
@@ -75,7 +74,6 @@ func (ni *Interface) Start(ctx context.Context) error {
 		ocpath.Root().InterfaceAny().Subinterface(0).Ipv6().AddressAny().PrefixLength().Config().PathStruct(),
```
[the remark this answers is not in the document]
Comment (DanG100) [reply dataplane/handlers/interface.go:74]: not sure added todo

File: dataplane/internal/engine/table.go [hunk 6]
```diff
@@ -218,6 +220,83 @@ func createFIBSelector(ctx context.Context, c fwdpb.ServiceClient) error {
...
+		},
+	}
+	if _, err := c.TableCreate(ctx, multicast); err != nil {
+		return err
+	}
+	return nil
+}
+
+// AddLayer2PuntRule adds rule to output packets to a corresponding port based on the destination MAC and input port.
+func AddLayer2PuntRule(ctx context.Context, c fwdpb.ServiceClient, nid uint64, mac, mask []byte) error {
```
[the remark this answers is not in the document]
Comment (DanG100) [reply dataplane/internal/engine/table.go:253]: renamed to nid to portID

File: dataplane/internal/engine/port.go [hunk 7]
```diff
@@ -74,14 +88,6 @@ func CreateExternalPort(ctx context.Context, c fwdpb.ServiceClient, name string)
 							},
 						},
 					}},
-					Outputs: []*fwdpb.ActionDesc{{ // Lookup in the to port to src MAC table.
-						ActionType: fwdpb.ActionType_ACTION_TYPE_LOOKUP,
-						Action: &fwdpb.ActionDesc_Lookup{
-							Lookup: &fwdpb.LookupActionDesc{
-								TableId: &fwdpb.TableId{ObjectId: &fwdpb.ObjectId{Id: srcMACTable}},
```
[the remark this answers is not in the document]
Comment (DanG100) [reply dataplane/internal/engine/port.go:81]: didn't mean to delete that, fixed

File: dataplane/internal/engine/port.go [hunk 8]
```diff
@@ -91,198 +97,3 @@ func CreateExternalPort(ctx context.Context, c fwdpb.ServiceClient, name string)
...
-					}},
-				},
-			},
-		},
-	}
-	if _, err := c.TableCreate(ctx, ipProtoPunt); err != nil {
-		return err
-	}
-
-	icmp := &fwdpb.TableEntryAddRequest{
```
[the remark this answers is not in the document]
Comment (DanG100) [reply dataplane/internal/engine/port.go:228]: added todo for layer3 punt, but any packet with dst_ip that is the configured ip of the port is punted to tap interface

File: dataplane/handlers/interface.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (wenovus) [reply dataplane/handlers/interface.go:341]: OK, I couldn't find good documentation on what this means but I'll trust your sources.

**Review by wenovus — APPROVED** [review]

**Pull request merged**